rfcs/0007-pass-spec.md

```diff
@@ -0,0 +1,463 @@
+
+By default, each of these passes returns immediately, unless pass-dependent invalidation flags are set or work is requested.
+Each pass can generally request work for later passes; for instance, the MUTATE pass can invalidate the layout of a widget, in which case the `layout` pass will run on that widget and its children and parents.
+
+Passes may also request work for *previous* passes, in which case all rewrite passes are run again in sequence.
+For instance, the UPDATE_POINTER may change a widget's size, requiring another layout pass.
+
+To avoid infinite loops in those cases, the number of reruns has a static limit.
+If passes are still requested past that limit, they're delayed to a later frame.
+
+#### MUTATE pass
+
+The **MUTATE** pass runs a list of callbacks with mutable access to the widget tree.
+These callbacks can be queued with the `mutate_later()` method of various context types.
+
+"Mutable access" means that those callbacks are given a `WidgetMut` to the widget that requested them, something that is otherwise only accessible from the owner of the global `RenderRoot` object (see "External Mutation" section).
+
+#### UPDATE_XXX passes
+
+Update passes mostly run internal calculations.
+They compute if some widget's property has changed, and send it a matching `on_update_status` event (see "Status" section below).
+
+For instance, if a user presses `Tab` and the event isn't handled in a widget, Masonry will run the `UPDATE_FOCUS` pass, which will automatically switch keyboard focus to the next focus-accepting widget.
+Both the previously-focused widget and the newly-focused widget will get an `on_update_status` call with relevant values.
+
+#### Layout pass
+
+The layout pass runs bidirectionally, passing constraints from the top down and getting back sizes and other layout info from the bottom up.
+
+It is subject to be reworked in the future to be closer to the semantics of web layout engines and the Taffy crate.
+
+Unlike with other passes, container widgets' `Widget::layout()` method must call `WidgetPod::layout()` on all of their children.
+Not doing so is a logical bug and will panic when debug assertions are on.
+
+#### Compose pass
+
+The **compose** pass runs top-down and assigns transforms to children.
+Transform-only layout changes (e.g. scrolling) should request compose instead of requesting layout.
+
+Compose is meant to be a cheaper way to position widgets than layout.
+Because the compose pass is more limited than layout, it's easier to recompute in many situations.
+
+For instance, if a widget in a list changes size, its siblings and parents must be re-laid out to account for the change; whereas changing a given widget's transform only affects its children.
+
+Masonry automatically calls the `compose` methods of all widgets in the tree, in depth-first preorder, where child order is determined by their position in the `children_ids()` array.
```

Comment:
I actually hope, we can design `layout` in a way that this is not really necessary (e.g. by dirty tracking via the flags and only recalculating the path that is affected by this). It would be really nice if we could (unlike in the web) make layout changes really interactive, e.g. data-driven animations etc. (which I think would play nice with Xilem) But I'm no expert with layouting algorithms, so I'm not really sure...

But I guess this is mostly meant to be a more general pass than `ParentWindowOrigin`

Reply:
Incremental layout is the goal, and there are ways to implement it, but it's not trivial. Some layout change *must* update sibling widgets.

For instance, if you have a widget defined as "be as wide as my widest widest child", if one of the children gets much wider, then all its siblings will have more available space and need to be re-laid out.

One of the things I'd like to implement is flags that deliberately rules out those cases, eg a flag a widget can set to say "my layout is independent from the layout of my children". Or "my height is only dependent on available width, no available height".

In any case, that's beyond this PR.